lib/search.js

```diff
@@ -0,0 +1,172 @@
+// code extracted and adapted from /front-end/platform/utilities.js
+// performSearchInContent is based on /front-end/common/ContentProvider.js implementation
+// _performSearchInContent tries to use regexp to gain performance
+
+
+var findAll = function(content, string)
+{
+  var matches = [];
+  var i = content.indexOf(string);
+  while (i !== -1) {
+    matches.push(i);
+    i = content.indexOf(string, i + string.length);
+  }
+  return matches;
+}
+
+var getlineEndings = function(content)
+{
+  var _lineEndings = findAll(content, "\n");
+  _lineEndings.push(content.length);
+  return _lineEndings;
+}
+
+var getlineAt = function(content, lineNumber, lineEndings) {
+  var lineStart = lineNumber > 0 ? lineEndings[lineNumber - 1] + 1 : 0;
+  var lineEnd = lineEndings[lineNumber];
+  var lineContent = content.substring(lineStart, lineEnd);
+  if (lineContent.length > 0 && lineContent.charAt(lineContent.length - 1) === "\r")
+    lineContent = lineContent.substring(0, lineContent.length - 1);
+  return lineContent;
+}
+
+var createSearchRegex = function(query, caseSensitive, isRegex)
+{
+  var regexFlags = caseSensitive ? "g" : "gi";
+  var regexObject;
+
+  if (isRegex) {
+    try {
+      regexObject = new RegExp(query, regexFlags);
+    } catch (e) {
+      // Silent catch.
+    }
+  }
+
+  if (!regexObject)
+    regexObject = createPlainTextSearchRegex(query, regexFlags);
+
+  return regexObject;
+}
+
+var createPlainTextSearchRegex = function(query, flags)
+{
+  // This should be kept the same as the one in ContentSearchUtils.cpp.
+  var regexSpecialCharacters = "^[]{}()\\.^$*+?|-,";
+  var regex = "";
+  for (var i = 0; i < query.length; ++i) {
+    var c = query.charAt(i);
+    if (regexSpecialCharacters.indexOf(c) != -1)
+      regex += "\\";
+    regex += c;
+  }
+  return new RegExp(regex, flags || "");
+}
+
+exports.performSearchInContent = function(content, query, caseSensitive, isRegex)
```

Comment:
@3y3 this code is adapted from /front-end/common/ContentProvider.js implementation...